File: hello-spark_2.3.0-java_1.8-maven/src/main/java/com/hello/spark/java/streaming/WordCount.java
package com.hello.spark.java.streaming;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.streaming.Durations;
import org.apache.spark.streaming.api.java.JavaDStream;
import org.apache.spark.streaming.api.java.JavaPairDStream;
import org.apache.spark.streaming.api.java.JavaReceiverInputDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import scala.Tuple2;

import java.util.Arrays;
import java.util.Iterator;

public class WordCount {

    public static void main(String[] args) throws InterruptedException {

        SparkConf conf = new SparkConf().setMaster("local[2]").setAppName("WordCountOnline");
        /**
         * 在创建streaminContext的时候 设置batch Interval
         */
        JavaStreamingContext jsc = new JavaStreamingContext(conf, Durations.seconds(5));


        JavaReceiverInputDStream<String> lines = jsc.socketTextStream("hadoop1", 9999);

        JavaDStream<String> words = lines.flatMap(new FlatMapFunction<String, String>() {
            /**
             *
             */
            private static final long serialVersionUID = 1L;

            @Override
            public Iterator<String> call(String line) throws Exception {
                return Arrays.asList(line.split(" ")).iterator();
            }
        });

        JavaPairDStream<String, Integer> pair = words.mapToPair(new PairFunction<String, String, Integer>() {
            /**
             *
             */
            private static final long serialVersionUID = 1L;

            @Override
            public Tuple2<String, Integer> call(String word) {
                return new Tuple2<String, Integer>(word, 1);
            }
        });

        JavaPairDStream<String, Integer> countResult = pair.reduceByKey(new Function2<Integer, Integer, Integer>() {
            /**
             *
             */
            private static final long serialVersionUID = 1L;

            @Override
            public Integer call(Integer i1, Integer i2) {
                return i1 + i2;
            }
        });

        countResult.print();

        jsc.start();
        jsc.awaitTermination();
 	    jsc.stop(false);

    }
}

File: hello-spark_2.3.1-scala_2.11.12-java_1.8-maven/src/main/scala/com/hello/spark/scala/streaming/WordCount.scala
package com.hello.spark.scala.streaming

import org.apache.spark.SparkConf
import org.apache.spark.storage.StorageLevel
import org.apache.spark.streaming.dstream.{DStream, ReceiverInputDStream}
import org.apache.spark.streaming.{Durations, StreamingContext}

object WordCount {
  def main(args: Array[String]): Unit = {
    val sparkConf : SparkConf = new SparkConf().setMaster("local[2]").setAppName("WordCount")


    val ssc : StreamingContext = new StreamingContext(sparkConf,Durations.seconds(5))

    ssc.checkpoint("d:/StreamingReceiverData")

    val linesDStream : ReceiverInputDStream[String] = ssc.socketTextStream("node", 9999, StorageLevel.MEMORY_AND_DISK_SER)

    val wordsDStream : DStream[String] = linesDStream.flatMap { _.split(" ") }

    // val pairDStream : DStream[Tuple2[String, Integer]] = wordsDStream.map { (_,1) }
    val pairDStream : DStream[(String, Integer)] = wordsDStream.map { (_,1) }

    val resultDStream : DStream[(String, Integer)] = pairDStream.reduceByKey(_+_)

    resultDStream.print()

    ssc.start()
    ssc.awaitTermination()
    ssc.stop()

  }
}
